ocaml-lsp-server/src/progress.ml

```diff
@@ -0,0 +1,78 @@
+open Import
+
+type enabled =
+  { mutable token : ProgressToken.t option
+  ; mutable build_counter : int
+  ; progress_report :
+      Server_notification.Progress.t ProgressParams.t -> unit Fiber.t
+  ; create_task : WorkDoneProgressCreateParams.t -> unit Fiber.t
+  }
+
+type t =
+  | Disabled
+  | Enabled of enabled
+
+let create (client_capabilities : ClientCapabilities.t) progress_report
+    create_task =
+  match client_capabilities.window with
+  | Some { workDoneProgress = Some true; _ } ->
+    Enabled { token = None; build_counter = 0; create_task; progress_report }
+  | _ -> Disabled
+
+let end_build (t : enabled) ~message =
+  match t.token with
+  | None -> Fiber.return ()
+  | Some token ->
+    t.token <- None;
+    t.progress_report
+      (ProgressParams.create ~token
+         ~value:
+           (Server_notification.Progress.End
+              (WorkDoneProgressEnd.create ~message ())))
+
+let end_build_if_running = function
+  | Disabled -> Fiber.return ()
+  | Enabled e -> end_build e ~message:"Build interrupted"
+
+let build_event t (event : Drpc.Build.Event.t) =
+  match t with
+  | Disabled -> Code_error.raise "progress reporting is not supported" []
+  | Enabled t -> (
+    match event with
+    | Finish -> end_build t ~message:"Build finished"
+    | Fail -> end_build t ~message:"Build failed"
+    | Interrupt -> end_build t ~message:"Build interrupted"
+    | Start ->
+      let open Fiber.O in
+      let* () = end_build t ~message:"Starting new build" in
+      let token = `String ("dune-build-" ^ Int.to_string t.build_counter) in
+      t.token <- Some token;
+      t.build_counter <- t.build_counter + 1;
+      t.create_task (WorkDoneProgressCreateParams.create ~token))
+
+let build_progress t (progress : Drpc.Progress.t) =
+  match t with
+  | Disabled -> Code_error.raise "progress reporting is not supported" []
```

Comment:
Raising seems odd to me: the `Enabled/Disabled` state is hidden from the user, so I would expect `build_progress` to do nothing if we are `Disabled`, instead of raising. This would also allow us to get rid of the check for `should_report_build_progress`

Reply:
It was like that before, but the checks were added to safe guard us from a very annoying bug: we are wasting a lot of cycles getting progress information from dune without making sure that we are going to do something productive with this information. I changed it to a programming error to make sure this doesn't happen.